gateway/src/apicast/policy/keycloak_scope_check/keycloak_scope_check.lua

```diff
@@ -0,0 +1,163 @@
+--- Keycloak Scope Check Policy
+-- This policy verifies the realm roles and the client roles in the JWT.
+--
+--- The realm roles are specified when you want to add scope check to the every client's resources or 3scale itself.
+-- https://www.keycloak.org/docs/4.0/server_admin/index.html#realm-roles
+--
+-- When you specify the realm roles in Keycloak, the JWT includes them as follows:
+--
+-- {
+--   "realm_access": {
+--     "roles": [
+--       "<realm_role_A>", "<realm_role_B>"
+--     ]
+--   }
+-- }
+--
+-- And you need to specify the "realm_roles" in this policy as follows:
+--
+-- "realm_roles": [
+--   { "name": "<realm_role_A>" }, { "name": "<realm_role_B>" }
+-- ]
+--
+--- The client roles are specified when you want to add scope check to the particular client's resources.
+-- https://www.keycloak.org/docs/4.0/server_admin/index.html#client-roles
+--
+-- When you specify the client roles in Keycloak, the JWT includes them as follows:
+--
+-- {
+--   "resource_access": {
+--     "<client_A>": {
+--       "roles": [
+--         "<client_role_A>", "<client_role_B>"
+--       ]
+--     },
+--     "<client_B>": {
+--       "roles": [
+--         "<client_role_A>", "<client_role_B>"
+--       ]
+--     }
+--   }
+-- }
+--
+-- And you need to specify the "client_roles" in this policy as follows:
+--
+-- "client_roles": [
+--   { "name": "<client_role_A>", "client": "<client_A>" },
```

Comment:
@y-tabata what are the implications of "client" being static in the configuration? Won't every client have there different id ? 
Won't you need this to be the actual client id from the JWT?

Reply:
The "client" means the property name of the json schema. The actual client is "<client_A>" and "<client_B>".

Reply:
Yes, but my point is that "client" value (client_A, client_B) will be dynamic, correct?

If for example you have a JWT from a Client "foobar" then this JWT might have `"resource_access": { "foobar": { "roles": ["finance"] }}`.
Now imagine you have 10000 clients. You don't want to maintain 10000 lines in the policy configuration saying that each client requires role finance. 

Will the client name in the "resource_access" refer to the "client id"  in the `azp`/`aud` claim?

Reply:
I think you raised an interesting question. 

A client can ask to use one of its client roles or another client role (see https://www.keycloak.org/docs/3.3/server_admin/topics/roles/client-roles.html) 

If you have two clients (app1, app2) with a role defined in each (app1 defines roleapp1, app2 defines roleapp2) and a user has the two roles, the access token would contain:
```
    "realm_access": {
    },
    "resource_access": {
        "app1": {
            "roles": [
                "roleapp1"
            ]
        },
        "app2": {
            "roles": [
                "roleapp2"
            ]
        }
    },
```

With this given configuration, there is not much room for interpretation: if I want to restrict access to GET /a to appa/roleappa and GET /b to appb/roleappb, the code is straightforward.

But if both apps define the same role ? This might be a perfectly valid expectation from the customer and especially with the PSD2 regulation. 

Given an API with two endpoints:
 - `GetAccountDetails`
 - `MoneyTransfer`

The bank might want to make sure:
1) the user is entitled to do bank transfers globally
2) the app has been granted to access the `MoneyTransfer` endpoint
3) the user is entitled to do a bank transfer on this specific application

In RH-SSO, this scenario is achieved as such:
1) is a realm role or client role of a well known client
2) is scope restriction (=> "Full scope allowed" is unchecked in the client config)
3) is a client role in each client that is given on a per user basis

In this configuration, the rules to access MoneyTransfer would be:
- the token has the realm role `money_transfer` (or the client role `bank/money_transfer`)
- the token has the client role `money_transfer` **of the current client**

So, I think the policy should allow both use cases: 
1. checking that the token issued for a client (`appA`) has a defined client role of a well known client (`appB/roleB`)
2. checking that the token issued for a client (`appX`) has a defined client role of itself (`appX/role`)

Suggestion on how to configure the policy for both use cases: 
1. `"client_roles": [ { "name": "roleB", "client": "appB" } ]`
2. `"client_roles": [ { "name": "role" } ]`

I tried to be as clear as possible. Let me know if not. 😄

Reply:
I'd like to have more explanation of the use case in the integration test for this.
IMO we can cover the 2nd point by https://github.com/3scale/apicast/pull/773#discussion_r198058559. Using liquid we can compare any value in the JWT to the roles.

But the first one is quite hard to do dynamically. How that would work with hundreds of clients @nmasse-itix  ? I suppose you don't want to configure this in the gateway for every new client.

Reply:
That's true, the first one is not meant to be dynamic. 

If you plan to use client roles this way, it should be well known clients roles, such as `realm-management/realm-admin` (for example to know if the user is administrator of the SSO Realm).

Right now, I cannot find a use case for referencing dynamically the role of **another** client. 

I agree that a design document would be helpful.

Reply:
@nmasse-itix I guess that with liquid support (https://github.com/3scale/apicast/pull/773#discussion_r198069574) we could use capture parameters from the mapping rules (or some other policy) as required role:


URL: `POST /MoneyTransfer/{client}`

```json
{
"client_roles": [
 { "name": "{{ usage.client }}", "name_type": "liquid", "client": "{{ usage.client }}", "client_type": "liquid" },
]
}
```

None of it is implemented yet, but at least we could have a path forward. 
@nmasse-itix do you think this information about another client would be available in the request ? URL, Headers, JWT, ...
If it would be somewhere in the request, we could match it and expose it for the scope verification.